**Business Understanding**

- Dataset source:  dataset from the MIMIC-III database
- Contains information on in-hospital mortality from the monitoring of patients in the Intensive Care Unit (ICU) for 48 hours

In [1]:
#loading required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Eye balling the data
df = pd.read_csv('ihm_48_hours.csv')
df.head()

Capillary refill rate  Diastolic blood pressure  Fraction inspired oxygen  \
0                    NaN                      73.0                       NaN   
1                    NaN                      73.0                       NaN   
2                    NaN                      73.0                       NaN   
3                    NaN                      73.0                       NaN   
4                    NaN                      73.0                       NaN   

  Glascow coma scale eye opening Glascow coma scale motor response  \
0                  Spontaneously                    Obeys Commands   
1                  Spontaneously                    Obeys Commands   
2                  Spontaneously                    Obeys Commands   
3                  Spontaneously                    Obeys Commands   
4                  Spontaneously                    Obeys Commands   

   Glascow coma scale total Glascow coma scale verbal response     Glucose  \
0                       NaN                           Oriented  -11.396037   
1                       NaN                           Oriented  115.000000   
2                       NaN                           Oriented  115.000000   
3                       NaN                           Oriented  115.000000   
4                       NaN                           Oriented  115.000000   

   Heart Rate  Height  Mean blood pressure  Oxygen saturation  \
0  -19.976803     NaN                 76.0          94.000000   
1   96.000000     NaN                 76.0          95.000000   
2   96.000000     NaN                 76.0          -6.497052   
3   96.000000     NaN                 76.0          95.000000   
4   96.000000     NaN                 76.0          95.000000   

   Respiratory rate  Systolic blood pressure  Temperature  Weight  pH  \
0              17.0                    116.0    36.388889    83.5 NaN   
1              18.0                    116.0    36.388889    83.5 NaN   
2              18.0                    116.0    36.388889    83.5 NaN   
3              18.0                    116.0    36.388889    83.5 NaN   
4              18.0                    116.0    36.388889    83.5 NaN   

  Patient_id  target  
0      30552       0  
1      30552       0  
2      30552       0  
3      30552       0  
4      30552       0

**Data Understanding**

| Column Name                         | Description                                                                 |
|-------------------------------------|-----------------------------------------------------------------------------|
| Capillary refill rate               | The time it takes for color to return to an external capillary bed (such as a fingertip) after pressure is applied. It is an indicator of peripheral perfusion. |
| Diastolic Blood Pressure            | The pressure in the arteries when the heart rests between beats. It's the lower of the two blood pressure measurements. |
| Fraction inspired oxygen            | The concentration of oxygen in the air mixture that is inhaled by the patient, often measured in ventilated patients. |
| Glascow coma scale eye opening      | A component of the Glasgow Coma Scale (GCS) that measures a patient's ability to open their eyes in response to stimuli. |
| Glascow coma scale motor response    | A component of the GCS that assesses a patient's motor response to stimuli, indicating brain function. |
| Glascow coma scale total            | The total score of the Glasgow Coma Scale, which assesses the level of consciousness in a person following a traumatic brain injury. |
| Glascow coma scale verbal response   | A component of the GCS that evaluates a patient’s verbal response, indicating their level of consciousness. |
| Glucose                             | The level of sugar (glucose) in the blood, an important indicator of metabolic status. |
| Heart Rate                          | The number of heartbeats per minute, indicating the functioning of the cardiovascular system. |
| Height                              | The measurement of a patient's stature, typically recorded in centimeters or meters. |
| Mean blood pressure                 | The average pressure in a patient's arteries during one cardiac cycle, indicating overall blood pressure. |
| Oxygen saturation                   | The percentage of hemoglobin binding sites in the bloodstream occupied by oxygen, indicating the oxygenation status of the patient. |
| Respiratory rate                    | The number of breaths taken per minute, indicating the patient’s respiratory status. |
| Systolic blood pressure             | The pressure in the arteries when the heart beats. It is the higher of the two blood pressure measurements. |
| Temperature                         | The body temperature of the patient, indicating metabolic and homeostatic status. |
| Weight                              | The body mass of the patient, typically recorded in kilograms. |
| pH                                  | The measure of acidity or alkalinity of the blood, indicating the balance of acids and bases in the body. |
| Patient_id                          | A unique identifier assigned to each patient. |
| target                              | The outcome or variable of interest- for this project -> mortality|

In [3]:
# Shape of the dataset
shape = df.shape
print('The dataset contains', shape[0], 'rows and', shape[1], 'columns.')

The dataset contains 300912 rows and 19 columns.


In [4]:
# checking the data types
df.dtypes

Capillary refill rate                 float64
Diastolic blood pressure              float64
Fraction inspired oxygen              float64
Glascow coma scale eye opening         object
Glascow coma scale motor response      object
Glascow coma scale total              float64
Glascow coma scale verbal response     object
Glucose                               float64
Heart Rate                            float64
Height                                float64
Mean blood pressure                   float64
Oxygen saturation                     float64
Respiratory rate                      float64
Systolic blood pressure               float64
Temperature                           float64
Weight                                float64
pH                                    float64
Patient_id                             object
target                                  int64
dtype: object

**Insights:**

- **Numerical variables** : Capillary refill rate ,Diastolic blood pressure, Fraction inspired oxygen, Glascow coma scale total, Glucose, Heart Rate, Height, Mean blood pressure, Oxygen saturation,Respiratory rate, Systolic blood pressure, Temperature, Weight, pH, target

- **Categorical variables** : Glascow coma scale eye opening, Glascow coma scale motor response, Glascow coma scale verbal response, Patient_id.




In [5]:
# Descriptive Statistics -> Categorical
df.describe(include='object')


Glascow coma scale eye opening Glascow coma scale motor response  \
count                          274190                            296978   
unique                              7                                12   
top                   4 Spontaneously                  6 Obeys Commands   
freq                            94516                            115595   

       Glascow coma scale verbal response Patient_id  
count                              296884     300912  
unique                                 12       6269  
top                          1.0 ET/Trach      30552  
freq                                87646         48

**Insights:**

- We do not have binary categorical variables present in the dataset, since none of the categorical variables have a unique value of 2.
- However we have nominal categorical variables(Have no order). They include:

- **1.Glascow coma scale eye opening**
    - This variable has 7 unique categories.
    - The category that has the highest number of observations is 4 Spontaneously which has 94516 observations.

- **2.Glascow coma scale motor response**
    - This variable has 12 unique categories.
    - The category that has the highest number of observations is 6 Obeys Commands which has 115595 observations.

- **3.Glascow coma scale verbal response**
    -This variable has 12 unique categoriries.
    - The category that has the highest number of observation is 1.0 ET/Trach which has 87646 observations.

- **4.Patient_id**
    - This variable has 6269 unique categories. This might imply 6269 patients participated in the study.
    - The count of this variable is 300912. Thus there is a possibility that data was collected from the patients more than once.


In [6]:
# Descriptive statistics -> Nuerical variables
df.describe()

Capillary refill rate  Diastolic blood pressure  \
count            6336.000000             296944.000000   
mean                0.219223                 62.541099   
std                 0.413753                341.559624   
min                 0.000000                  0.000000   
25%                 0.000000                 51.000000   
50%                 0.000000                 59.000000   
75%                 0.000000                 69.000000   
max                 1.000000             100105.010000   

       Fraction inspired oxygen  Glascow coma scale total        Glucose  \
count              88464.000000             184416.000000  300698.000000   
mean                   0.599884                 10.818123     130.628329   
std                    0.253919                  4.334923      84.171126   
min                    0.000000                  3.000000     -19.999974   
25%                    0.400000                  8.000000     101.000000   
50%                    0.500000                 11.000000     126.000000   
75%                    0.700000                 15.000000     158.000000   
max                    7.100000                 15.000000    9999.000000   

          Heart Rate        Height  Mean blood pressure  Oxygen saturation  \
count  300912.000000  55824.000000         296984.00000      300912.000000   
mean       79.447793    168.543422             78.79108          95.343460   
std        32.145920     15.137414             29.52986        2529.203751   
min       -19.999623      0.000000            -34.00000         -19.999687   
25%        70.000000    160.000000             68.00000          95.000000   
50%        84.000000    170.000000             77.00000          98.000000   
75%        97.000000    178.000000             88.00000         100.000000   
max       941.000000    203.000000           9381.00000      981023.000000   

       Respiratory rate  Systolic blood pressure    Temperature  \
count     300864.000000            300912.000000  298848.000000   
mean          18.731265               119.694213      36.832834   
std            6.884248                23.396042       1.000075   
min            0.000000                 0.000000       0.000000   
25%           15.000000               103.000000      36.277802   
50%           18.000000               117.000000      36.833333   
75%           22.000000               134.000000      37.388889   
max         1211.000000               295.000000      73.760002   

              Weight             pH         target  
count  221040.000000  230614.000000  300912.000000  
mean       82.969018       5.573617       0.142128  
std        26.765857       5.963634       0.349182  
min         0.000000     -19.999706       0.000000  
25%        66.600000       7.310000       0.000000  
50%        79.099998       7.370000       0.000000  
75%        94.699997       7.420000       0.000000  
max       931.224376      99.000000       1.000000

**Insights:**

- **Capillary refill rate:**   
    - Has a 6336 data points
    - Has a range from 0 to 1, hence implying it might be a binary variable.
- **Diastolic blood pressure:**
    - Has 296944 data points recorded.
    - High standard deviation(Mean = 62.541, standard deviation = 341.559).
    - Ranges from 0 to 100105. Compared to the quartiles, this variable might have outliers.
    
- **Fraction inspired oxygen:**
    - The mean is 0.599,with a standard deviation of 0.25
    - Ranges from 0 to 7.1. 
- **Glascow coma scale total:**
    - The mean is 10.818,with a standard deviation of 4.33
    - Ranges from 3 to 15, with most values clustering around the higher end of this range. 

- **Glucose:**
    - The mean glucose level is 130.63 ,with a standard deviation of 84.17, which is quite large
    - Ranges from 0 to -19.99 to 9999.   

- **Heart Rate:**
    - Has 300912 records
    - Most rates average at around a mean of 79 beats per minute
    - Has a standard deviation of of 32.15 (moderate variability)
    - The least value is -19.999623 and maximum value is 941
    - based on the quatiles, most values fall between 70 and 97 beats per minutes(bpm) with a median value of 84 bpm
    - Note that there might exist errors in the data collected especially during data entry 

- **Height:**
    - Has 55824 records
    - Most heights average at around a mean of 168.54
    - Has a standard deviation of 15.14 which is not a very big deviation from the central points 
    - Most of the heights fall between 160 to 178 with a median of 170
    - Tallest individual had a height of 203 and shortest individual had a height of 0 (probable error here)

- **Mean blood pressure:**
    - Has 296984 records
    - Most levels average at around a mean of 78.79 mg/dL.
    - Has a standard deviation of 29.529 which is exudes a deviation from the central points 
    - Most of the levels fall between 68 and 88 with
    - Data set has a minimum value of -34 and max value of 9381, this is a possible data error

- **Oxygen saturation:**
    - Has 300912 records
    - Data averages with a mean of 95.343460
    - Has a standard deviation of 2529.203752, which is very high. This shows that there may exist outliers in the data
    - The Oxygen saturation levels fall between 95 and 100 with a central point of 98
    - The least saturation level recorded is -19.999687 and highest value of 981023(this is a possible error)

- **Respiratory rate:**
    - Has 300864 records
    - Data averages with a mean of 18.731265
    - Has as standard deviation of 6.884248(moderate variability), which is significantly ok hence data quality might be ok
    - The lowest value is 0 and largest being 1211(this may be erronoeus since the value is quite large)
    - The rates range between 15 and 22 with a median of 18

- **Systolic blood pressure:**
    - Has  300912 records
    - Data averages with a mean of  119.694213
    - Has as standard deviation of 23.396042(moderate variablity)
    - The lowest value is 0 and largest being 295
    - The rates range between 103 and 134 with a median of 117

- **Temperature:**
    - Has 298848 records
    - Data averages around 36.832834°C
    - Has as standard deviation of 1.000, which is small hence data quality might be ok
    - The lowest level is 0°C and largest being 73.76°C
    - The rates range between 36.277802°C and 37.388889°C with a median of 36.833333°C

- **Weight:**
    - Has 221040 records
    - Data averages with a mean of 82.969018
    - Has a standard deviation of 26.765857, which expresses that the data quality might not be ok
    - The lowest weight is at 0 and largest at 931.224376
    - The weights range between 66.600000 and 94.699997 with a median value of 79.099998

- **pH:**
    - Has 230614 records
    - Data averages with a mean of 5.573617
    - Has a standard deviation of 5.963634, which is not as high but expresses that the data quality might not be ok but at an insignificant level
    - The lowest weight is at -19.999706 and largest at 99
    - The weights range between 7.310000 and 7.420000 with a median value of 7.370000, the data seems to be compromised in term sof data quality because of the min and max values there exists outliers 



*Generally:*
- *Based on the above the data quality is challenged, as some features contain outliers and other contain missing values.*
              


In [9]:
# checking the percentage of missing values
missing_percentage = df.isnull().mean() * 100
print(missing_percentage)


Capillary refill rate                 97.894401
Diastolic blood pressure               1.318658
Fraction inspired oxygen              70.601372
Glascow coma scale eye opening         8.880337
Glascow coma scale motor response      1.307359
Glascow coma scale total              38.714309
Glascow coma scale verbal response     1.338597
Glucose                                0.071117
Heart Rate                             0.000000
Height                                81.448397
Mean blood pressure                    1.305365
Oxygen saturation                      0.000000
Respiratory rate                       0.015952
Systolic blood pressure                0.000000
Temperature                            0.685915
Weight                                26.543308
pH                                    23.361647
Patient_id                             0.000000
target                                 0.000000
dtype: float64


**Insights:**
- The following features have more than 50% of missing values: Capillary refill rate(97.89%), fraction inspired oxygen(70.60%), Height(81.44%).
-  The following features have less than 50% of missing values: Diastolic blood pressure(1.318),Glascow coma scale eye opening,Glascow coma scale motor response, Glascow coma scale total  .
- 